### PR TITLE
[WIP] Cythonizes GROParser & GROReader very minimally

### PR DESCRIPTION
**Pull request by Fenil3510**
Changes made in this Pull Request:
 -  Cythonized version of GRO class is called when parsing ascii file. Not sure if this is the best way to do it. Please suggest changes.

PR Checklist
------------
 - [ ] Tests?
 - [ ] Docs?
 - [ ] CHANGELOG updated?
 - [ ] Issue raised/referenced?

### CODE REVIEW

**Comment by orbeckst** [comment]
In [your email](https://groups.google.com/d/msg/mdnalysis-discussion/YFXD_ZLzIVg/a_2TrqyGAwAJ) you said
> I have made changes in the GROParser and created a cythonized version(very basic changes). It gives an improvement of 0.05 seconds on average on a 2mb gro file. This is not much given that cythonization is not extensively done. 

The code does not seem to run on the CI (AppVeyor fails) but taking for a moment that you got your changes to build locally, it is interesting that the simplest approach of just wrapping the code in cython does not change much. 

Do you know where the Python code spends most of its time, i.e., what is the performance bottleneck? Did you profile the Python GRO parser (e.g. with [line_profiler](https://github.com/rkern/line_profiler)) to find out? This will give you a better idea of what needs optimizing.

(Also, eventually the PR needs to build on the CI and we need to figure out if we can move the cython code to the lib module - might be tricky for parsers because code in lib is supposed to be independent of MDAnalysis things such as topologies and AtomGroups. But first explore and see if there are some obvious ways to speed up the parser.)

**Comment by orbeckst** [comment]
@Fenil3510 , I remembered that a while ago PR #1862 was started with the goal to leverage the [chemfiles python interface](http://chemfiles.org/chemfiles.py/0.8.0/) for some trajectory formats. They don't support all formats but a [a fair number](http://chemfiles.org/chemfiles/latest/formats.html#list-of-supported-formats), at least as coordinate "Readers/Writers". Not sure if they can be used as topology "parsers", too. 

chemfiles contains a [GRO reader/writer](http://chemfiles.org/chemfiles/latest/devdoc/format.html?highlight=gro#_CPPv4N9chemfiles3GROE) (written in C/C++ but apparently based on VMD's [molfile plugin](http://www.ks.uiuc.edu/Research/vmd/plugins/molfile/)). I don't know how seamlessly chemfiles can interact with MDAnalysis.

We didn't have this in our project description, partly because the PR stalled, partly because it is complicated enough to understand MDAnalysis and we don't necessarily want you to having to spend time learning another library that we ourselves know little about. However, I feel that this should be mentioned so that you have a better idea of what else is going on in MDAnalysis, as potentially related to your project.

If you or anyone else has opinions, please discuss them here.

cc @richardjgowers @jbarnoud

**Comment by Fenil3510** [comment]
> e CI and we need to figure out if we can move the cython code to the lib module

Thanks for the suggestions @orbeckst. Yes, I just wrapped the code in Cython. I'll see why tests are failing. I'll look into line_profiler to find out what's to be done.

**Comment by richardjgowers** [comment]
@orbeckst I think the problem with adopting C++/C for reading ascii is it then becomes very hard (for python devs) to patch.  Currently our stuff bends over backwards to read everything because we can patch it.

@Fenil3510 I think this might be a good place to get ideas for how to do quick ascii i/o:

http://wesmckinney.com/blog/speeding-up-pandass-file-parsers-with-cython/

Then obviously finding the corresponding cython code in the pandas repo (if they still use Cython).

**Comment by Luthaf** [comment]
Hey, chemfiles dev here! So PR #1862 have stalled a bit, but I just finished the release of version 0.9 of chemfiles so I'll try to back to it. It was basically working, just needed a bit of documentation. It might take a bit of time as I am currently writing my thesis ...

> Not sure if they can be used as topology "parsers", too.

It should be easy to convert from chemfiles Topology class to MDAnalysis version. Current version of #1862 only uses the coordinates and throw away the topology, but I could update it to include chemfiles as a topology reader. 

> chemfiles contains a GRO reader/writer (written in C/C++ but apparently based on VMD's molfile plugin). I don't know how seamlessly chemfiles can interact with MDAnalysis.

We actually have our own implementation of GRO format, not based on VMD molfiles. It lives here: https://github.com/chemfiles/chemfiles/blob/master/src/formats/GRO.cpp.

> I think the problem with adopting C++/C for reading ascii is it then becomes very hard (for python devs) to patch. Currently our stuff bends over backwards to read everything because we can patch it.

We try to write readable and easy to modify C++ code in chemfiles, and modern C++ is closer to Python than old school C. But there is always a barrier to overcome when contributing to a project in a different language.

File: package/MDAnalysis/topology/GROParser.py [hunk 1]
```diff
@@ -63,97 +63,4 @@
...
-            Segids(np.array(['SYSTEM'], dtype=object))
-        ]
-
-        top = Topology(n_atoms=n_atoms, n_res=len(new_resids), n_seg=1,
-                       attrs=attrs,
-                       atom_resindex=residx,
-                       residue_segindex=None)
-
-        return top
+from .c_GROParser import GROParser
```
Comment (tylerjereddy) [inline package/MDAnalysis/topology/GROParser.py]: You need to build the C extension into the library via a `setup.py` file so that this doesn't cause issues during the build, which is also visible in CI.

For examples of MDA compiled extension building -- `git grep 'MDAExtension'`. I don't think we normally commit the Cython generated code to our repo. Probably best to just remove that file from version control & set up proper building of the extension.

**Comment by tylerjereddy** [comment]
>  finding the corresponding cython code in the pandas repo

+1 to at least check this; `pandas` goes to great lengths to optimize some if its file loaders and trying to optimize text file reading can be a pretty deep rabbit hole.

File: package/MDAnalysis/topology/GROParser.py (hunk 1 above)
Comment (Fenil3510) [reply package/MDAnalysis/topology/GROParser.py]: Oh, I did build it with setup.py but didn't know how to commit it right. Probably that might have caused the build error. The `.so` files are ignored anyway. So ones can you clarify what files to include after `setup.py`. I'm building with `python setup.py build_ext --inplace`.

Comment (Fenil3510) [reply package/MDAnalysis/topology/GROParser.py]: Ok, I get it. So I need to make an `MDAExtension` object for it in the `setup.py` file and cythonize list it there itself, right?

Comment (tylerjereddy) [reply package/MDAnalysis/topology/GROParser.py]: Yup, CI is all green now, so I'll mark this resolved

**Comment by Fenil3510** [comment]
@tylerjereddy @orbeckst. I profiled the code of GROParser line by line and checked what takes the most time. It was obviously the looping part, so I thought to make the for loop cythonized. I read [here](https://cython-docs2.readthedocs.io/en/latest/src/tutorial/numpy.html) that cythons access to numpy array is much faster for indexing than the regular array. I made the changes accordingly, but I get an improvement of only 40ms on average, ie. actual average time reduces from 710 to 640ms. I still have to dig into `pandas` parsers as suggested. Any other obvious cythonization I'm missing?

File: package/MDAnalysis/topology/c_GROParser.pyx [hunk 2]
```diff
@@ -86,6 +87,12 @@ class GROParser(TopologyReaderBase):
         # resid, resname, name, index, (x,y,z)
         cdef int n_atoms;
         cdef int i;
+        cdef str line
+        cdef int count = 0
+        cdef np.ndarray[np.int32_t] resids
```
Comment (Fenil3510) [inline package/MDAnalysis/topology/c_GROParser.pyx]: Defining arrays with numpy  C interface is causing the build to fail probably, though the build is successful locally. Any ideas how to deal with this?

File: package/MDAnalysis/topology/c_GROParser.pyx [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+        # Gro has the following columns
+        # resid, resname, name, index, (x,y,z)
+        cdef int n_atoms;
+        cdef int i;
+        cdef str line
+        cdef int count = 0
+        cdef np.ndarray[np.int32_t] resids
+        cdef np.ndarray[object] resnames
+        cdef np.ndarray[object] names
+        cdef np.ndarray[np.int32_t] indices
```
Comment (tylerjereddy) [inline package/MDAnalysis/topology/c_GROParser.pyx]: better to use [typed memoryviews](https://cython.readthedocs.io/en/latest/src/userguide/memoryviews.html) here vs. the old buffer syntax

File: package/MDAnalysis/topology/c_GROParser.pyx [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+        # Fix wrapping of resids (if we ever saw a wrap)
+        if np.any(resids == 0):
+            # find places where resid hit zero again
+            wraps = np.where(resids == 0)[0]
+            # group these places together:
+            # find indices of first 0 in each block of zeroes
+            # 1) find large changes in index, (ie non sequential blocks)
+            diff = np.diff(wraps) != 1
+            # 2) make array of where 0-blocks start
+            starts = np.hstack([wraps[0], wraps[1:][diff]])
```
Comment (tylerjereddy) [inline package/MDAnalysis/topology/c_GROParser.pyx]: `starts` is initialized here without prior type declaration

**Comment by Fenil3510** [comment]
@tylerjereddy numpy functions `np.any` and `np.where` don't seem to work with cython typed memoryviews. Would have to write custom function if they are to be used.

**Comment by tylerjereddy** [comment]
You can use constructs like `np.asarray(memoryview)` where needed, but if there's an opportunity for creative speedup with custom loops you may want to look into it. Memoryviews usually can work without the GIL unlike the buffer interface, so it may require some careful design to really take advantage.

**Comment by tylerjereddy** [comment]
NumPy also has a C API

**Comment by Fenil3510** [comment]
@tylerjereddy , I see. I will change it to `memoryview` and see if the custom function has any improvement. I was focussing more on the for loop `for line in inf` and that's where most of the time goes (according to line_profiler). So I was thinking of
1. Faster indexing of numpy array (which I expected to speed up after using buffer interface, but didn't)
2. Custom type conversion function(no python interaction) to replace `int(line[0:5])` but it ended being rather costly
3. Can string slicing be improved in any way?

I'd be grateful for any suggestions.

**Comment by Fenil3510** [comment]
The current changes in the `GROParser` ie. `parse` function to be specific. Profiling the `parse` function separately gives an improvement of 2.5 - 3x on any file (files tested 2MB to 17 MB). But the overall reading from `mda.Universe()` shows a negligible gain. This is because parsing takes 6% of the total time. So this is surely not the bottleneck. I chose `GROParser` for getting started. I will be looking into `GROReader` for some time and probably find out the bottleneck
Any suggestions are welcomed.

**Comment by Fenil3510** [comment]
@orbeckst as you said
> Also, eventually the PR needs to build on the CI and we need to figure out if we can move the cython code to the lib module - might be tricky for parsers because code in lib is supposed to be independent of MDAnalysis things such as topologies and AtomGroups.
I was wondering what is the reason to move cython code to `lib` module? Sorry, if the question is very basic.

**Comment by orbeckst** [comment]
When you look at [`lib.formats`](https://github.com/MDAnalysis/mdanalysis/tree/develop/package/MDAnalysis/lib/formats) you see that these _readers_ are written in a way independent from core MDAnalysis code (i.e., no universes, AtomGroups, Topology, etc). In #1194 we aimed for making more of our *readers* low-level and MDAnalysis-agnostic. When I said

> we need to figure out if we can move the cython code to the lib module - might be tricky for parsers because code in lib is supposed to be independent of MDAnalysis things such as topologies and AtomGroups.

I was expressing the desire to do the same thing for _parsers_ but also expressed that I kind of doubted that this could be done easily.

In short: Moving GROParser to `lib.formats` is not high priority. I was interested in knowing if it was even a possibility.

File: package/MDAnalysis/topology/GROParser.py (hunk 1 above)
Comment (orbeckst) [inline package/MDAnalysis/topology/GROParser.py]: I am not an expert in cython extensions so I have to ask this as a comment here: Why keep a stub Python file instead of naming the cython file just GROParser.pyx and building it as `topology.GROParser`?

cc @tylerjereddy @richardjgowers

**Comment by Fenil3510** [comment]
@orbeckst thankyou for the information. Yes, the intermediate Python file can be removed I believe, I had kept it to play safe, now that I've familiarized with Cython modules, looks like it shouldn't cause an issue, I'll make the changes soon.

**Comment by Fenil3510** [comment]
Had some tests failing in `test_nuclinfo.py`, so had to merge develop into this.

**Comment by Fenil3510** [comment]
Hmm, test fails due to a numpy merge [here](https://github.com/numpy/numpy/pull/13359/files). Some changes in pickling.

File: package/MDAnalysis/topology/GROParser.pyx [hunk 5]
```diff
@@ -47,6 +47,7 @@
 from __future__ import absolute_import
 
 import numpy as np
+cimport numpy as np
```
Comment (richardjgowers) [inline package/MDAnalysis/topology/GROParser.pyx]: this is only needed when you have numpy on the left hand side of things, eg `cdef np.ndarray[float] arr`.  You're using memoryviews here (which is better), so you just need the Python import to get numpy functions imported.

File: package/MDAnalysis/topology/GROParser.pyx [hunk 6]
```diff
@@ -84,6 +85,16 @@ def parse(self, **kwargs):
...
         # resid, resname, name, index, (x,y,z)
+        cdef int n_atoms;
+        cdef int i;
+        cdef str line
+        cdef int count = 0
+        cdef int[:] resids
+        cdef object[:] resnames
+        cdef object[:] names
+        cdef int[:] indices
+        cdef long long[:] starts
```
Comment (richardjgowers) [inline package/MDAnalysis/topology/GROParser.pyx]: pretty sure we just need ints here too

**Comment by richardjgowers** [comment]
@Fenil3510 don't worry about those failures.  What are loading times like with these changes?

File: package/MDAnalysis/topology/GROParser.pyx (hunk 5 above)
Comment (Fenil3510) [reply package/MDAnalysis/topology/GROParser.pyx]: Yes, my bad. Forgot to remove after I shifted to memoryviews.

File: package/MDAnalysis/topology/GROParser.pyx (hunk 6 above)
Comment (Fenil3510) [reply package/MDAnalysis/topology/GROParser.pyx]: Travis was failing with `int` and expecting it to be `long long` as I remember. I'll double check.

Comment (Fenil3510) [reply package/MDAnalysis/topology/GROParser.pyx]: `np.where` returns a np.int64 but mac integer is 32-bit as the same issure [here](https://stackoverflow.com/questions/32262976/cython-buffer-type-mismatch-expected-int-but-got-long), I've type-casted the arrays inplace with `np.int32` and works with `int` now.

**Comment by orbeckst** [comment]
@Fenil3510 could you please benchmark the parser on its own and full universe creation with the GRO files in the vesicle data set https://www.mdanalysis.org/MDAnalysisData/vesicles.html ?
```
conda install -c conda-forge MDAnalysisData
```
and 
```python
from MDAnalyisData.datasets import fetch_vesicle_lib

dataset = fetch_vesicle_lib()

grofiles = dataset.structure    # should be 3 GRO files
```

These are fairly big systems (see `dataset.labels` and the web page for description). It would ne nice to have a reproducible benchmark that we can also report, e.g., in a blog post or paper.

**Comment by Fenil3510** [comment]
@orbeckst sorry for the delay, have my exams running parallelly. 
Here are the results and benchmarks
1.  This is only the parser (ie. parse function to be specific), shows gains of 2.5x to 3.5x
benchmark_1

2. Gains in making the whole universe are less for smallest file but upto 1.5x on biggest file. This is expected though.
image

Working on cythonizing the readers will definitely help increase gains.

**Comment by orbeckst** [comment]
@Fenil3510 many thanks, very useful and encouraging! 

Good luck with your exams!

**Comment by Fenil3510** [comment]
> It would be nice to have a reproducible benchmark that we can also report, e.g., in a blog post or paper.
I will start working on this after May 9th.

**Comment by orbeckst** [comment]
I merged current develop, hopefully the tests will all pass then.

**Comment by Fenil3510** [comment]
I've been busy lately, I'll get back on this pretty soon.

**Comment by orbeckst** [comment]
Not sure why the old numpy test (**NUMPY_VERSION=1.10.4**) fails, [Job #1030.5](https://travis-ci.com/MDAnalysis/mdanalysis/jobs/205747332)

```
==================================== ERRORS ====================================
_________________ ERROR collecting MDAnalysisTests/test_api.py _________________
ImportError while importing test module '/home/travis/build/MDAnalysis/mdanalysis/testsuite/MDAnalysisTests/test_api.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
testsuite/MDAnalysisTests/test_api.py:29: in <module>
    import MDAnalysis as mda
package/MDAnalysis/__init__.py:184: in <module>
    from .lib import log
package/MDAnalysis/lib/__init__.py:42: in <module>
    from . import pkdtree
package/MDAnalysis/lib/pkdtree.py:35: in <module>
    from scipy.spatial import cKDTree
../../../miniconda/envs/test/lib/python3.5/site-packages/scipy/spatial/__init__.py:99: in <module>
    from .kdtree import *
../../../miniconda/envs/test/lib/python3.5/site-packages/scipy/spatial/kdtree.py:8: in <module>
    import scipy.sparse
../../../miniconda/envs/test/lib/python3.5/site-packages/scipy/sparse/__init__.py:230: in <module>
    from .csr import *
../../../miniconda/envs/test/lib/python3.5/site-packages/scipy/sparse/csr.py:13: in <module>
    from ._sparsetools import (csr_tocsc, csr_tobsr, csr_count_blocks,
E   ImportError: numpy.core.multiarray failed to import
------------------------------- Captured stderr --------------------------------
RuntimeError: module compiled against API version 0xb but this version of numpy is 0xa
```

I am restarting the job... maybe it's a conda glitch???

**Comment by Fenil3510** [comment]
Looks like an issue with Travis. All other PR's are failing the same test.

**Comment by Fenil3510** [comment]
I'd need some help here. I used types in GROReader for some variables and converted it into Cython file(This method worked with GROParser). But too many tests are failing for this commit, while I can build the universe with `MDAnalysis.datasets` with this commit and gains 3-4x gains but tests are failing as I can't access position method of `Atom` class which is weird. I'm unable to trace the exact reason for this, any help would be much appreciated.

**Comment by richardjgowers** [comment]
Hey Fenil   
  

  
Thanks for sticking with this. I think it might be the reader isn’t getting registered when the module imports... can you import mdanalysis and check mda._READERS (or whatever that dict is called....)
  

  
  

  
  
>

File: package/MDAnalysis/coordinates/GRO.pyx [hunk 7]
```diff
@@ -116,6 +116,10 @@
 from ..core import flags
 from ..exceptions import NoDataError
 from ..lib import util
+cimport cython
+@cython.boundscheck(False)  # Deactivate bounds checking
```
Comment (richardjgowers) [inline package/MDAnalysis/coordinates/GRO.pyx]: these decorators aren't doing what you are expecting here, they won't apply to the whole file.  I think they're probably just modifying the Timestep object.

You probably need to decorate the class method directly

**Comment by richardjgowers** [comment]
@Fenil3510 the reader class wasn't getting added to `mda._READERS` which is a dict of all the reader classes which is usually done automatically by the class it inherits from... Not sure why, but I've manually added it for now in `coordinates.__init__`.  I did a rebase and force pushed to your branch, but it's "working" now, in that the code is ran and throws an error :)

**Comment by Fenil3510** [comment]
Yeah, it's failing silently even though there is an except to raise `KeyError` if it's not in the registry. Looking into it.

**Comment by richardjgowers** [comment]
hey @Fenil3510 

I played around with writing a Cython XYZReader function here:

https://gist.github.com/richardjgowers/dbec308c364aae8b9ca84e8e22171823

There's no error checking (this would need to be added!) but it seems to go pretty fast if you want some ideas.
